dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java

```diff
@@ -0,0 +1,581 @@
+ * called during:
+ * 1) Worker shutdown or close as part of EtcdMembershipManager close
+ * 2) FileSystemContext closeContext as part of EtcdMembershipManager close
+ * As we never set mClient to be null after connect, also jetcd client can be closed idempotently
+ * so it's ok to ignore thread safety for close()
+ *
+ * As for jetcd Client, it's managing its own connect/reconnect/loadbalance to other etcd
+ * instances, will leave these logic to jetcd client itself for now unless we need to
+ * handle it in our layer.
+ */
+public class AlluxioEtcdClient implements Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioEtcdClient.class);
+  private static final Lock INSTANCE_LOCK = new ReentrantLock();
+  public static final long DEFAULT_LEASE_TTL_IN_SEC = 2L;
+  public static final long DEFAULT_TIMEOUT_IN_SEC = 2L;
+  public static final int RETRY_TIMES = 3;
+  private static final int RETRY_SLEEP_IN_MS = 100;
+  private static final int MAX_RETRY_SLEEP_IN_MS = 500;
+  @GuardedBy("INSTANCE_LOCK")
+  @Nullable
+  private static volatile AlluxioEtcdClient sAlluxioEtcdClient;
+  public final ServiceDiscoveryRecipe mServiceDiscovery;
+  private final AtomicBoolean mConnected = new AtomicBoolean(false);
+  private final Closer mCloser = Closer.create();
+  // only watch for children change(add/remove) for given parent path
+  private final ConcurrentHashMap<String, Watch.Watcher> mRegisteredWatchers =
+      new ConcurrentHashMap<>();
+  private Client mClient;
+  private final String[] mEndpoints;
+
+  /**
+   * CTOR for AlluxioEtcdClient.
+   * @param conf
+   */
+  public AlluxioEtcdClient(AlluxioConfiguration conf) {
+    String clusterName = conf.getString(PropertyKey.ALLUXIO_CLUSTER_NAME);
+    List<String> endpointsList = conf.getList(PropertyKey.ETCD_ENDPOINTS);
+    mEndpoints = endpointsList.toArray(new String[0]);
+    mServiceDiscovery = new ServiceDiscoveryRecipe(this, clusterName);
+  }
+
+  /**
+   * Get the singleton instance of AlluxioEtcdClient.
+   * @param conf
+   * @return AlluxioEtcdClient
+   */
+  public static AlluxioEtcdClient getInstance(AlluxioConfiguration conf) {
+    if (sAlluxioEtcdClient == null) {
+      try (LockResource lockResource = new LockResource(INSTANCE_LOCK)) {
+        if (sAlluxioEtcdClient == null) {
+          sAlluxioEtcdClient = new AlluxioEtcdClient(conf);
+        }
+      }
+    }
+    return sAlluxioEtcdClient;
+  }
+
+  /**
+   * Create jetcd grpc client no forcing.
+   */
+  public void connect() {
+    connect(false);
+  }
+
+  /**
+   * Create jetcd grpc client and force(or not) connection.
+   * @param force
+   */
+  public void connect(boolean force) {
+    if (mConnected.get() && !force) {
+      return;
+    }
+    mConnected.set(false);
+    // create client using endpoints
+    Client client = Client.builder().endpoints(mEndpoints)
+        .build();
+    if (mConnected.compareAndSet(false, true)) {
```

Comment:
there is a conner case, when the two clients are created at the same time, and first one can be assigned to mClient, the second one would be left without close.

Reply:
fixed in #17890